c7n/actions/notify.py

```diff
@@ -14,9 +14,97 @@
 from c7n.version import version
 
 
+class ResourceMessageBuffer:
+
+    # conservative ratio calculated over all extant json test data
+    # files, most resources have many common repeated keys and values
+    # re compress down well.
+    #
+    # base64 increases size, but the compression still reduces total size versus raw.
+    # https://lemire.me/blog/2019/01/30/what-is-the-space-overhead-of-base64-encoding/
+    #
+    # script to caculate ratio
+    # https://gist.github.com/kapilt/8c3558a7db0d178cb1c4e91d47dacc77
+    #
+    # we use this conservative value as a seed and adapt based on observed data
+    b64_zlib_ratio = 0.6
+
+    def __init__(self, envelope, buffer_max_size):
+        self.buffer_max_size = buffer_max_size
+        self.resource_parts = []
+
+        envelope['resources'] = []
+        self.envelope = utils.dumps(envelope)
+        self.raw_size = float(len(self.envelope))
+        self.observed_ratio = 0
+
+    def add(self, resource):
+        self.resource_parts.append(utils.dumps(resource))
+        self.raw_size += len(self.resource_parts[0])
```

Comment:
Shouldn't this be `self.raw_size += len(self.resource_parts[-1])` if we want to add the raw_size of the last appended resource? Else, it would keep adding the size of first resource over and over again. It also seems to be fixing the issue. Please find the logs below:
```
2023-07-20 21:30:04,095: custodian.cache:DEBUG Disabling cache
2023-07-20 21:30:04,096: custodian.commands:DEBUG Loaded file ecs.yml. Contains 1 policies
2023-07-20 21:30:04,283: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:04,628: custodian.output:DEBUG Storing output with <LogFile file://output/ecs-service/custodian-run.log>
2023-07-20 21:30:04,743: custodian.policy:DEBUG Running policy:ecs-service resource:ecs-service region:us-east-1 c7n:0.9.29
2023-07-20 21:30:04,839: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:05,360: custodian.resources.ecscluster:DEBUG Filtered from 424 to 424 ecscluster
2023-07-20 21:30:33,248: custodian.resources.service:DEBUG Filtered from 567 to 567 service
2023-07-20 21:30:33,248: custodian.policy:INFO policy:ecs-service resource:ecs-service region:us-east-1 count:567 time:28.50
2023-07-20 21:30:33,930: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:34,263: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:7
2023-07-20 21:30:34,522: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:34,882: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:40
2023-07-20 21:30:35,162: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:35,576: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:41
2023-07-20 21:30:35,843: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:36,230: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:36
2023-07-20 21:30:36,517: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:36,878: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:41
2023-07-20 21:30:37,158: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:37,573: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:27
2023-07-20 21:30:37,863: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:38,312: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:38
2023-07-20 21:30:38,587: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:38,928: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:40
2023-07-20 21:30:39,211: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:39,606: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:26
2023-07-20 21:30:39,884: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:40,243: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:36
2023-07-20 21:30:40,524: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:40,865: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:40
2023-07-20 21:30:41,131: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:41,545: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:46
2023-07-20 21:30:41,818: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:42,162: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:47
2023-07-20 21:30:42,429: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:42,841: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:48
2023-07-20 21:30:43,110: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:43,454: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:35
2023-07-20 21:30:43,724: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:30:44,074: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:19
2023-07-20 21:30:44,074: custodian.policy:INFO policy:ecs-service action:notify resources:567 execution_time:10.52
2023-07-20 21:30:44,076: custodian.output:DEBUG metric:ResourceCount Count:567 policy:ecs-service restype:ecs-service scope:policy
2023-07-20 21:30:44,076: custodian.output:DEBUG metric:ApiCalls Count:720 policy:ecs-service restype:ecs-service
```

It also works without the multipler (*3). Seems to be faster without the multiplier. 
```2023-07-20 21:38:59,146: custodian.cache:DEBUG Disabling cache
2023-07-20 21:38:59,146: custodian.commands:DEBUG Loaded file ecs.yml. Contains 1 policies
2023-07-20 21:38:59,338: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:38:59,721: custodian.output:DEBUG Storing output with <LogFile file://output/ecs-service/custodian-run.log>
2023-07-20 21:38:59,831: custodian.policy:DEBUG Running policy:ecs-service resource:ecs-service region:us-east-1 c7n:0.9.29
2023-07-20 21:38:59,926: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:00,545: custodian.resources.ecscluster:DEBUG Filtered from 424 to 424 ecscluster
2023-07-20 21:39:27,361: custodian.resources.service:DEBUG Filtered from 567 to 567 service
2023-07-20 21:39:27,361: custodian.policy:INFO policy:ecs-service resource:ecs-service region:us-east-1 count:567 time:27.53
2023-07-20 21:39:28,057: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:28,474: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:9
2023-07-20 21:39:28,748: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:29,069: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:43
2023-07-20 21:39:29,342: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:29,779: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:43
2023-07-20 21:39:30,006: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:30,393: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:38
2023-07-20 21:39:30,672: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:31,070: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:42
2023-07-20 21:39:31,355: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:31,682: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:34
2023-07-20 21:39:31,971: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:32,323: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:34
2023-07-20 21:39:32,613: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:32,992: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:44
2023-07-20 21:39:33,266: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:33,613: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:31
2023-07-20 21:39:33,897: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:34,303: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:41
2023-07-20 21:39:34,590: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:34,916: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:50
2023-07-20 21:39:35,195: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:35,539: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:49
2023-07-20 21:39:35,814: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:36,204: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:43
2023-07-20 21:39:36,476: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:36,923: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:43
2023-07-20 21:39:37,134: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-07-20 21:39:37,472: custodian.actions:INFO sent message:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX policy:ecs-service template:default count:23
2023-07-20 21:39:37,472: custodian.policy:INFO policy:ecs-service action:notify resources:567 execution_time:9.81
2023-07-20 21:39:37,474: custodian.output:DEBUG metric:ResourceCount Count:567 policy:ecs-service restype:ecs-service scope:policy
2023-07-20 21:39:37,474: custodian.output:DEBUG metric:ApiCalls Count:719 policy:ecs-service restype:ecs-service
```

Reply:
yeah that's a bug

Reply:
fixed